### PR TITLE
Adds a failing test for #3060

### LINKED ISSUE #3060: `hypothesis@6.14.8` stopped working for `dry-python/returns`
Dependabot PR, that updates version `6.14.7` to `6.14.8`, revealed that hypothesis has stopped working for us: https://github.com/dry-python/returns/pull/1028/checks?check_run_id=3351483158

Previous version did work fine: https://github.com/dry-python/returns/commit/50f06330c00c5c8a5538b192bb73bed109a052a0

Logs:
```
You can reproduce this example by temporarily adding @reproduce_failure('6.14.8', b'AA==') as a decorator on your test case
_______________________ test_pair_altablen_identity_law ________________________

    def factory(source: st.DataObject) -> None:
>       with type_vars():

returns/contrib/hypothesis/laws.py:204: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.venv/lib/python3.9/site-packages/hypothesis/core.py:624: in execute_once
    result = self.test_runner(data, run)
.venv/lib/python3.9/site-packages/hypothesis/executors.py:52: in default_new_style_executor
    return function(data)
.venv/lib/python3.9/site-packages/hypothesis/core.py:620: in run
    return test(*args, **kwargs)
returns/contrib/hypothesis/laws.py:204: in factory
    with type_vars():
.venv/lib/python3.9/site-packages/hypothesis/core.py:562: in test
    result = self.test(*args, **kwargs)
returns/contrib/hypothesis/laws.py:207: in factory
    source.draw(st.builds(law.definition))
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/core.py:1639: in draw
    result = self.conjecture_data.draw(strategy)
.venv/lib/python3.9/site-packages/hypothesis/internal/conjecture/data.py:871: in draw
    strategy.validate()
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/strategies.py:400: in validate
    self.do_validate()
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/lazy.py:135: in do_validate
    w.validate()
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/core.py:792: in validate
    fixed_dictionaries(self.kwargs).validate()
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/strategies.py:400: in validate
    self.do_validate()
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/lazy.py:135: in do_validate
    w.validate()
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/strategies.py:400: in validate
    self.do_validate()
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/strategies.py:800: in do_validate
    self.mapped_strategy.validate()
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/strategies.py:400: in validate
    self.do_validate()
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/collections.py:46: in do_validate
    s.validate()
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/strategies.py:400: in validate
    self.do_validate()
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/lazy.py:135: in do_validate
    w.validate()
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/strategies.py:401: in validate
    self.is_empty
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/strategies.py:135: in accept
    recur(self)
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/strategies.py:131: in recur
    mapping[strat] = getattr(strat, calculation)(recur)
.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/deferred.py:70: in calc_is_empty
    return recur(self.wrapped_strategy)
            return st.from_type(annotated_type)
        # Now, confirm that we're dealing with a generic type as we expected
        if sys.version_info[:2] < (3, 9) and not isinstance(
            thing, typing_root_type
        ):  # pragma: no cover
            raise ResolutionFailed(f"Cannot resolve {thing} to a strategy")
    
        # Some "generic" classes are not generic *in* anything - for example both
        # Hashable and Sized have `__args__ == ()` on Python 3.7 or later.
        # (In 3.6 they're just aliases for the collections.abc classes)
        origin = getattr(thing, "__origin__", thing)
        if (
            typing.Hashable is not collections.abc.Hashable
            and origin in vars(collections.abc).values()
            and len(getattr(thing, "__args__", None) or []) == 0
        ):
            return st.from_type(origin)
    
        # Parametrised generic types have their __origin__ attribute set to the
        # un-parametrised version, which we need to use in the subclass checks.
        # e.g.:     typing.List[int].__origin__ == typing.List
        mapping = {
            k: v
            for k, v in _global_type_lookup.items()
            if is_generic_type(k) and try_issubclass(k, thing)
        }
        if typing.Dict in mapping or typing.Set in mapping:
            # ItemsView can cause test_lookup.py::test_specialised_collection_types
            # to fail, due to weird isinstance behaviour around the elements.
            mapping.pop(typing.ItemsView, None)
            if sys.version_info[:2] == (3, 6):  # pragma: no cover
                # `isinstance(dict().values(), Container) is False` on py36 only -_-
                mapping.pop(typing.ValuesView, None)
        if typing.Deque in mapping and len(mapping) > 1:
            # Resolving generic sequences to include a deque is more trouble for e.g.
            # the ghostwriter than it's worth, via undefined names in the repr.
            mapping.pop(typing.Deque)
        if len(mapping) > 1:
            # issubclass treats bytestring as a kind of sequence, which it is,
            # but treating it as such breaks everything else when it is presumed
            # to be a generic sequence or container that could hold any item.
            # Except for sequences of integers, or unions which include integer!
            # See https://github.com/HypothesisWorks/hypothesis/issues/2257
            #
            # This block drops ByteString from the types that can be generated
            # if there is more than one allowed type, and the element type is
            # not either `int` or a Union with `int` as one of its elements.
            elem_type = (getattr(thing, "__args__", None) or ["not int"])[0]
            if getattr(elem_type, "__origin__", None) is typing.Union:
                union_elems = elem_type.__args__
else:
                union_elems = ()
            if not any(
                isinstance(T, type) and issubclass(int, T)
                for T in list(union_elems) + [elem_type]
            ):
                mapping.pop(typing.ByteString, None)
        elif (
            (not mapping)
            and isinstance(thing, ForwardRef)
            and thing.__forward_arg__ in vars(builtins)
        ):
            return st.from_type(getattr(builtins, thing.__forward_arg__))
        strategies = [
            v if isinstance(v, st.SearchStrategy) else v(thing)
            for k, v in mapping.items()
            if sum(try_issubclass(k, T) for T in mapping) == 1
        ]
        empty = ", ".join(repr(s) for s in strategies if s.is_empty)
        if empty or not strategies:
>           raise ResolutionFailed(
                f"Could not resolve {empty or thing} to a strategy; "
                "consider using register_type_strategy"
            )
E           hypothesis.errors.ResolutionFailed: Could not resolve returns.interfaces.altable.AltableN[~_FirstType, ~_SecondType, ~_ThirdType] to a strategy; consider using register_type_strategy

.venv/lib/python3.9/site-packages/hypothesis/strategies/_internal/types.py:334: ResolutionFailed
---------------------------------- Hypothesis ----------------------------------
Falsifying example: factory(
    source=data(...),
)
```

### PR DESCRIPTION
**Pull request by sobolevn**
It is surely not the easiest reproduction, but I've tried to simplify it as much as I could (`returns` is the most complicated project in the respect of type-annotations that I am aware of).

Here's the root of the problem:
- When I put `print(thing, thing_args, thing_orig_bases)` in `try_issubclass`, this is the result: `<class 'returns.interfaces.mappable.MappableN'> None (typing.Generic[~_FirstType, ~_SecondType, ~_ThirdType], returns.primitives.laws.Lawful[ForwardRef('MappableN[_FirstType, _SecondType, _ThirdType]')])`
- Then, it tries to resolve `returns.primitives.laws.Lawful[ForwardRef('MappableN[_FirstType, _SecondType, _ThirdType]')`
- But, its arg is a string, not a `TypeVar`
- Everything collapses 💥 

Fixes #3060.

### CODE REVIEW

**Comment by sobolevn** [comment]
I will work on simplification once more, I guess I can remove a lot of details.

**Comment by sobolevn** [comment]
Looks like a minimal example to me 👀

**Comment by Zac-HD** [comment]
> But, its arg is a string, not a TypeVar

Aha! I bet this is the problem - if you change try_issubclass to check if the args are typevars *or strings*, does that fix the issue? I'll check in the morning if you can't get to it.

**Comment by sobolevn** [comment]
One more thing: this exception https://github.com/HypothesisWorks/hypothesis/blob/25a1a48d9494efc52b44370c46591ccba3f2aba3/hypothesis-python/src/hypothesis/strategies/_internal/types.py#L112-L116 get eaten because it is a subtype of `TypeError`

I will add a special case for it.

**Comment by sobolevn** [comment]
I am going to leave this to you, @Zac-HD 
Because I don't understand the reasons behind the last diff, I don't want to make it worse 😨

**Comment by Zac-HD** [comment]
Thanks @sobolevn - that made it really easy to identify the problem... I'd literally hard-coded an error if the thing-to-resolve inherited from more than one parametrised generic type.

Can you run the `returns` test suite against this patch, or as much as is practical, to check that it works for the full complexity of everything you're doing?  Once that passes I'll merge a fix and "yank" 6.14.8 so that nobody else gets it.

**Comment by sobolevn** [comment]
It is better now! However, there are still some failures:

```
______________ test_requirescontextresult_readerresultbasedn_purity_law ______________

    def factory(source: st.DataObject) -> None:
>       with type_vars():

returns/contrib/hypothesis/laws.py:214: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.venv/lib/python3.7/site-packages/hypothesis/core.py:624: in execute_once
    result = self.test_runner(data, run)
.venv/lib/python3.7/site-packages/hypothesis/executors.py:52: in default_new_style_executor
    return function(data)
.venv/lib/python3.7/site-packages/hypothesis/core.py:620: in run
    return test(*args, **kwargs)
returns/contrib/hypothesis/laws.py:214: in factory
    with type_vars():
.venv/lib/python3.7/site-packages/hypothesis/core.py:562: in test
    result = self.test(*args, **kwargs)
returns/contrib/hypothesis/laws.py:217: in factory
    source.draw(st.builds(law.definition))
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/core.py:1639: in draw
    result = self.conjecture_data.draw(strategy)
.venv/lib/python3.7/site-packages/hypothesis/internal/conjecture/data.py:871: in draw
    strategy.validate()
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/strategies.py:400: in validate
    self.do_validate()
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/lazy.py:135: in do_validate
    w.validate()
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/core.py:792: in validate
    fixed_dictionaries(self.kwargs).validate()
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/strategies.py:400: in validate
    self.do_validate()
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/lazy.py:135: in do_validate
    w.validate()
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/strategies.py:400: in validate
    self.do_validate()
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/strategies.py:800: in do_validate
    self.mapped_strategy.validate()
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/strategies.py:400: in validate
    self.do_validate()
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/collections.py:46: in do_validate
    s.validate()
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/strategies.py:400: in validate
    self.do_validate()
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/lazy.py:135: in do_validate
    w.validate()
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/strategies.py:401: in validate
    self.is_empty
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/strategies.py:135: in accept
    recur(self)
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/strategies.py:131: in recur
    mapping[strat] = getattr(strat, calculation)(recur)
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/deferred.py:70: in calc_is_empty
    return recur(self.wrapped_strategy)
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/deferred.py:40: in wrapped_strategy
    result = self.__definition()
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/core.py:944: in <lambda>
    lambda thing: deferred(lambda: _from_type(thing)),
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/core.py:1062: in _from_type
    return types.from_typing_type(thing)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

thing = returns.interfaces.specific.reader_result.ReaderResultBasedN[~_FirstType, ~_SecondType, ~_ThirdType]

    def from_typing_type(thing):
        # We start with special-case support for Union and Tuple - the latter
        # isn't actually a generic type. Then we handle Literal since it doesn't
        # support `isinstance`.
        #
        # We then explicitly error on non-Generic types, which don't carry enough
        # information to sensibly resolve to strategies at runtime.
        # Finally, we run a variation of the subclass lookup in `st.from_type`
        # among generic types in the lookup.
        if getattr(thing, "__origin__", None) == tuple or isinstance(
            thing, getattr(typing, "TupleMeta", ())
        ):
            elem_types = getattr(thing, "__tuple_params__", None) or ()
            elem_types += getattr(thing, "__args__", None) or ()
            if (
                getattr(thing, "__tuple_use_ellipsis__", False)
                or len(elem_types) == 2
                and elem_types[-1] is Ellipsis
            ):
                return st.lists(st.from_type(elem_types[0])).map(tuple)
            elif len(elem_types) == 1 and elem_types[0] == ():
                return st.tuples()  # Empty tuple; see issue #1583
            return st.tuples(*map(st.from_type, elem_types))
        if hasattr(typing, "Final") and getattr(thing, "__origin__", None) == typing.Final:
            return st.one_of([st.from_type(t) for t in thing.__args__])
        if is_typing_literal(thing):
            args_dfs_stack = list(thing.__args__)
            literals = []
            while args_dfs_stack:
                arg = args_dfs_stack.pop()
                if is_typing_literal(arg):
                    args_dfs_stack.extend(reversed(arg.__args__))
                else:
                    literals.append(arg)
            return st.sampled_from(literals)
        if is_annotated_type(thing):  # pragma: no cover
            # This requires Python 3.9+ or the typing_extensions package
            args = thing.__args__
            assert args, "it's impossible to make an annotated type with no args"
            annotated_type = args[0]
            return st.from_type(annotated_type)
        # Now, confirm that we're dealing with a generic type as we expected
        if sys.version_info[:2] < (3, 9) and not isinstance(
            thing, typing_root_type
        ):  # pragma: no cover
            raise ResolutionFailed(f"Cannot resolve {thing} to a strategy")
    
        # Some "generic" classes are not generic *in* anything - for example both
        # Hashable and Sized have `__args__ == ()` on Python 3.7 or later.
        # (In 3.6 they're just aliases for the collections.abc classes)
        origin = getattr(thing, "__origin__", thing)
        if (
            typing.Hashable is not collections.abc.Hashable
            and origin in vars(collections.abc).values()
            and len(getattr(thing, "__args__", None) or []) == 0
        ):
            return st.from_type(origin)
    
        # Parametrised generic types have their __origin__ attribute set to the
        # un-parametrised version, which we need to use in the subclass checks.
        # e.g.:     typing.List[int].__origin__ == typing.List
        mapping = {
            k: v
            for k, v in _global_type_lookup.items()
            if is_generic_type(k) and try_issubclass(k, thing)
        }
        if typing.Dict in mapping or typing.Set in mapping:
            # ItemsView can cause test_lookup.py::test_specialised_collection_types
            # to fail, due to weird isinstance behaviour around the elements.
            mapping.pop(typing.ItemsView, None)
            if sys.version_info[:2] == (3, 6):  # pragma: no cover
                # `isinstance(dict().values(), Container) is False` on py36 only -_-
                mapping.pop(typing.ValuesView, None)
        if typing.Deque in mapping and len(mapping) > 1:
            # Resolving generic sequences to include a deque is more trouble for e.g.
            # the ghostwriter than it's worth, via undefined names in the repr.
            mapping.pop(typing.Deque)
        if len(mapping) > 1:
            # issubclass treats bytestring as a kind of sequence, which it is,
            # but treating it as such breaks everything else when it is presumed
            # to be a generic sequence or container that could hold any item.
            # Except for sequences of integers, or unions which include integer!
            # See https://github.com/HypothesisWorks/hypothesis/issues/2257
            #
            # This block drops ByteString from the types that can be generated
            # if there is more than one allowed type, and the element type is
            # not either `int` or a Union with `int` as one of its elements.
            elem_type = (getattr(thing, "__args__", None) or ["not int"])[0]
            if getattr(elem_type, "__origin__", None) is typing.Union:
                union_elems = elem_type.__args__
            else:
                union_elems = ()
            if not any(
                isinstance(T, type) and issubclass(int, T)
                for T in list(union_elems) + [elem_type]
            ):
                mapping.pop(typing.ByteString, None)
        elif (
            (not mapping)
            and isinstance(thing, ForwardRef)
            and thing.__forward_arg__ in vars(builtins)
        ):
            return st.from_type(getattr(builtins, thing.__forward_arg__))
        strategies = [
            v if isinstance(v, st.SearchStrategy) else v(thing)
            for k, v in mapping.items()
            if sum(try_issubclass(k, T) for T in mapping) == 1
        ]
        empty = ", ".join(repr(s) for s in strategies if s.is_empty)
        if empty or not strategies:
            raise ResolutionFailed(
>               f"Could not resolve {empty or thing} to a strategy; "
                "consider using register_type_strategy"
            )
E           hypothesis.errors.ResolutionFailed: Could not resolve returns.interfaces.specific.reader_result.ReaderResultBasedN[~_FirstType, ~_SecondType, ~_ThirdType] to a strategy; consider using register_type_strategy

.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/types.py:336: ResolutionFailed
```

I will try to figure out what wrong.

File: hypothesis-python/src/hypothesis/strategies/_internal/types.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/strategies/_internal/types.py]: I don't think this is correct - if we have multiple bases I think we want to check that the args are compatible for every base with args, not just that there is at least one base with compatible args.

Comment (sobolevn) [reply hypothesis-python/src/hypothesis/strategies/_internal/types.py]: Ok, I see. Here's the code that does not work for me with this change: https://github.com/dry-python/returns/blob/1e7eeee2816dd8525e8596a83dc700648410025d/returns/interfaces/specific/reader_result.py#L102-L113

As you can see this class has 3 different parents with different number of type arguments and their type arguments are incompatible with each other on the first sight. However, it type-checks perfectly.

**Comment by sobolevn** [comment]
I hope I didn't get over-excited with the tests 😆

File: hypothesis-python/src/hypothesis/strategies/_internal/types.py
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/strategies/_internal/types.py]: Ah, makes sense. I'll re-review properly in the morning, but I think with this we're ready to merge?

Comment (sobolevn) [reply hypothesis-python/src/hypothesis/strategies/_internal/types.py]: Yes, I think so too. Thanks a lot for your help! 👍

**Review by Zac-HD — APPROVED** [review]
LGTM - thanks @sobolevn for your report, and assistance in testing and solving the bug 😍

**Comment by sobolevn** [comment]
It worked: https://github.com/dry-python/returns/pull/1033 🎉

**Comment by Zac-HD** [comment]
Hurray!

**Pull request merged**
